File: desktop/src/main/drivers/tizen.js
import { EventEmitter } from "node:events"
import WebSocket from "ws"
import { KEYS as SHARED, plain } from "../shared.js"

/**
 * Samsung Tizen (2016+) — a JSON WebSocket on 8002.
 *
 * On the first connection the TV shows an "allow this device?" dialog and, once
 * accepted, hands back a token in its ms.channel.connect frame. The token is
 * stored and appended to the URL from then on.
 *
 * Written directly against `ws`: the published wrappers still pull in the
 * deprecated `request` package.
 */

const KEYS = plain(SHARED.tizen)

const b64 = (s) => Buffer.from(s, "utf8").toString("base64")

/** Read the unauthenticated info endpoint — also the cheapest liveness check. */
export async function fetchDeviceInfo(host, { timeout = 3000 } = {}) {
  const res = await fetch(`http://${host}:8001/api/v2/`, { signal: AbortSignal.timeout(timeout) })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  const info = await res.json()
  return {
    name: info?.device?.name ?? null,
    model: info?.device?.modelName ?? null,
    mac: info?.device?.wifiMac ?? null,
    powered: info?.device?.PowerState ? info.device.PowerState === "on" : null,
    tokenAuth: info?.device?.TokenAuthSupport === "true",
  }
}

export class TizenDriver extends EventEmitter {
  static kind = "tizen"

  constructor(device) {
    super()
    this.device = device
    this.ws = null
    this.retry = null
    this.stopped = false
    this.status = { connected: false, powered: null, volume: null, muted: null, currentApp: null }
  }

  get capabilities() {
    return [...Object.keys(KEYS), "text", "launch"]
  }

  #update(patch) {
    this.status = { ...this.status, ...patch }
    this.emit("status", this.status)
  }

  async connect() {
    this.stopped = false
    this.#clearRetry()
    if (this.ws) await this.disconnect({ keepRetry: true })

    const params = new URLSearchParams({ name: b64("Mac Remote") })
    if (this.device.creds?.token) params.set("token", this.device.creds.token)

    const url = `wss://${this.device.host}:8002/api/v2/channels/samsung.remote.control?${params}`
    // The TV presents a self-signed certificate it will not let you replace,
    // so chain validation is off; the connection stays inside the LAN.
    const ws = new WebSocket(url, { rejectUnauthorized: false, handshakeTimeout: 8000 })
    this.ws = ws

    if (!this.device.creds?.token) this.emit("pair-required", { kind: "confirm" })

    ws.on("message", (raw) => {
      let msg
      try {
        msg = JSON.parse(raw.toString())
      } catch {
        return
      }
      if (msg.event === "ms.channel.connect") {
        this.#update({ connected: true, powered: true })
        if (msg.data?.token && msg.data.token !== this.device.creds?.token) {
          this.emit("paired", { token: String(msg.data.token) })
        }
      } else if (msg.event === "ms.channel.unauthorized") {
        this.emit("error", new Error("הטלוויזיה דחתה את הבקשה — יש לאשר את המכשיר בתפריט"))
      }
    })

    ws.on("close", () => {
      this.#update({ connected: false })
      this.#scheduleRetry()
    })
    ws.on("error", (err) => {
      this.emit("error", err)
      this.#update({ connected: false, powered: false })
    })

    await new Promise((resolve, reject) => {
      ws.once("open", resolve)
      ws.once("error", reject)
    })
  }

  #send(payload) {
    if (this.ws?.readyState !== WebSocket.OPEN) throw new Error("הטלוויזיה אינה מחוברת")
    this.ws.send(JSON.stringify(payload))
  }

  async send(cmd, arg) {
    if (cmd === "text") {
      return this.#send({
        method: "ms.remote.control",
        params: { Cmd: b64(String(arg ?? "")), DataOfCmd: "base64", TypeOfRemote: "SendInputString" },
      })
    }
    if (cmd === "launch") {
      return this.#send({
        method: "ms.channel.emit",
        params: { event: "ed.apps.launch", to: "host", data: { appId: String(arg), action_type: "DEEP_LINK" } },
      })
    }

    const key = KEYS[cmd]
    if (!key) throw new Error(`פקודה לא נתמכת: ${cmd}`)
    return this.#send({
      method: "ms.remote.control",
      params: { Cmd: "Click", DataOfCmd: key, Option: "false", TypeOfRemote: "SendRemoteKey" },
    })
  }

  async wake() {
    const { wake } = await import("../wol.js")
    if (!this.device.mac) throw new Error("אין כתובת MAC — יש להדליק ידנית פעם אחת כדי לזהות אותה")
    await wake(this.device.mac)
    // Give the panel a moment to bring its network stack up before reconnecting.
    setTimeout(() => this.connect().catch(() => {}), 4000)
  }

  #scheduleRetry() {
    if (this.retry || this.stopped) return
    this.retry = setTimeout(() => {
      this.retry = null
      this.connect().catch(() => {})
    }, 15_000)
  }

  #clearRetry() {
    if (this.retry) clearTimeout(this.retry)
    this.retry = null
  }

  async disconnect({ keepRetry = false } = {}) {
    if (!keepRetry) {
      this.stopped = true
      this.#clearRetry()
    }
    try {
      this.ws?.removeAllListeners()
      this.ws?.close()
    } catch {}
    this.ws = null
    this.#update({ connected: false })
  }
}

File: mobile/lib/src/protocol/tizen/tizen.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

/// Samsung Tizen (2016 and later) — JSON over a WebSocket on 8002.
///
/// On the first connection the television shows an "allow this device?" prompt
/// and, once accepted, returns a token in its `ms.channel.connect` frame. The
/// token is replayed on every later connection, so the prompt appears once.
///
/// The set presents a self-signed certificate it will not let you replace, so
/// chain validation is turned off. The connection never leaves the home network.
class TizenClient {
  TizenClient({
    required this.host,
    this.token,
    this.name = 'Orbit',
    this.port = 8002,
    this.timeout = const Duration(seconds: 10),
  });

  final String host;
  final int port;
  final String name;
  final Duration timeout;

  /// Returned by the television after the prompt is accepted; null on first
  /// contact, which is what makes the prompt appear.
  String? token;

  WebSocket? _socket;

  final _tokens = StreamController<String>.broadcast();
  final _errors = StreamController<Object>.broadcast();
  final _closed = StreamController<void>.broadcast();

  /// Emits once when the television hands back a fresh token to store.
  Stream<String> get tokens => _tokens.stream;
  Stream<Object> get errors => _errors.stream;
  Stream<void> get closed => _closed.stream;

  bool get isConnected => _socket != null;

  static String _b64(String value) => base64.encode(utf8.encode(value));

  /// The unauthenticated info endpoint — also the cheapest liveness check, and
  /// the only way to learn the MAC address needed to wake the set later.
  static Future<TizenInfo?> fetchInfo(
    String host, {
    Duration timeout = const Duration(seconds: 3),
  }) async {
    final client = HttpClient()..connectionTimeout = timeout;
    try {
      final request = await client.getUrl(
        Uri.parse('http://$host:8001/api/v2/'),
      );
      final response = await request.close().timeout(timeout);
      if (response.statusCode != 200) return null;
      final json = jsonDecode(
        await response.transform(utf8.decoder).join(),
      ) as Map<String, dynamic>;
      final device = json['device'] as Map<String, dynamic>? ?? const {};
      return TizenInfo(
        name: device['name'] as String?,
        model: device['modelName'] as String?,
        mac: device['wifiMac'] as String?,
        powered: switch (device['PowerState']) {
          'on' => true,
          final String _ => false,
          _ => null,
        },
      );
    } on Object {
      return null;
    } finally {
      client.close();
    }
  }

  Future<void> connect() async {
    await close();

    final query = <String, String>{
      'name': _b64(name),
      if (token != null && token!.isNotEmpty) 'token': token!,
    };
    final uri = Uri(
      scheme: 'wss',
      host: host,
      port: port,
      path: '/api/v2/channels/samsung.remote.control',
      queryParameters: query,
    );

    final client = HttpClient()
      ..connectionTimeout = timeout
      ..badCertificateCallback = (_, _, _) => true;

    final socket = await WebSocket.connect(
      uri.toString(),
      customClient: client,
    ).timeout(timeout);
    _socket = socket;

    socket.listen(
      _onMessage,
      onDone: () {
        _socket = null;
        if (!_closed.isClosed) _closed.add(null);
      },
      onError: (Object error) {
        if (!_errors.isClosed) _errors.add(error);
      },
      cancelOnError: false,
    );
  }

  void _onMessage(dynamic raw) {
    Map<String, dynamic> message;
    try {
      message = jsonDecode('$raw') as Map<String, dynamic>;
    } on Object {
      return;
    }

    switch (message['event']) {
      case 'ms.channel.connect':
        final fresh = (message['data'] as Map?)?['token'];
        if (fresh != null && '$fresh' != token) {
          token = '$fresh';
          if (!_tokens.isClosed) _tokens.add(token!);
        }
      case 'ms.channel.unauthorized':
        if (!_errors.isClosed) {
          _errors.add(
            const TizenException(
              'הטלוויזיה דחתה את הבקשה — יש לאשר את המכשיר על המסך',
            ),
          );
        }
    }
  }

  void _send(Map<String, dynamic> payload) {
    final socket = _socket;
    if (socket == null) throw const TizenException('הטלוויזיה אינה מחוברת');
    try {
      socket.add(jsonEncode(payload));
    } on Object catch (failure) {
      // A socket the television reset asynchronously throws on write; letting
      // it escape would take down the zone rather than the one command.
      throw TizenException('$failure');
    }
  }

  /// Press a remote key, by the name in the shared table (e.g. KEY_VOLUP).
  void sendKey(String key) => _send({
    'method': 'ms.remote.control',
    'params': {
      'Cmd': 'Click',
      'DataOfCmd': key,
      'Option': 'false',
      'TypeOfRemote': 'SendRemoteKey',
    },
  });

  void sendText(String text) => _send({
    'method': 'ms.remote.control',
    'params': {
      'Cmd': _b64(text),
      'DataOfCmd': 'base64',
      'TypeOfRemote': 'SendInputString',
    },
  });

  void launch(String appId) => _send({
    'method': 'ms.channel.emit',
    'params': {
      'event': 'ed.apps.launch',
      'to': 'host',
      'data': {'appId': appId, 'action_type': 'DEEP_LINK'},
    },
  });

  Future<void> close() async {
    final socket = _socket;
    _socket = null;
    try {
      await socket?.close();
    } on Object {
      // Already gone.
    }
  }

  Future<void> dispose() async {
    await close();
    await _tokens.close();
    await _errors.close();
    await _closed.close();
  }
}

class TizenInfo {
  const TizenInfo({this.name, this.model, this.mac, this.powered});
  final String? name;
  final String? model;
  final String? mac;
  final bool? powered;
}

class TizenException implements Exception {
  const TizenException(this.message);
  final String message;
  @override
  String toString() => message;
}
